language/bytecode_verifier/src/instantiation_loops.rs

```diff
@@ -0,0 +1,202 @@
+    /// //
+    /// //    edge: foo_T -> Id -> bar_T
+    Identity,
+    /// This type of edge from type formal T1 to T2 means T2 is instantiated with a type resulted
+    /// by applying one or more type constructors to T1 (potentially with other types).
+    ///
+    /// This is interesting to us as it creates a new (and bigger) type.
+    ///
+    /// Example:
+    /// //    struct Baz<T> {}
+    /// //    foo<T>() { bar<Baz<T>>(); return; }
+    /// //
+    /// //    edge: foo_T -> TyConApp -> bar_T
+    TyConApp(&'a SignatureToken),
+}
+
+pub struct InstantiationLoopChecker<'a> {
+    module: &'a CompiledModule,
+
+    graph: Graph<Node, Edge<'a>>,
+    node_map: HashMap<Node, NodeIndex>,
+}
+
+impl<'a> InstantiationLoopChecker<'a> {
+    pub fn new(module: &'a CompiledModule) -> Self {
+        Self {
+            module,
+            graph: Graph::new(),
+            node_map: HashMap::new(),
+        }
+    }
+
+    /// Retrives the node corresponding to the specified type formal.
+    /// If none exists in the graph yet, create one.
+    fn get_or_add_node(&mut self, node: Node) -> NodeIndex {
+        match self.node_map.entry(node) {
+            hash_map::Entry::Occupied(entry) => *entry.get(),
+            hash_map::Entry::Vacant(entry) => {
+                let idx = self.graph.add_node(node);
+                entry.insert(idx);
+                idx
+            }
+        }
+    }
+
+    /// Helper function that extracts type parameters from a given type.
+    /// Duplicated entries are removed.
+    fn extract_type_parameters(ty: &SignatureToken) -> Vec<TypeParameterIndex> {
+        use SignatureToken::*;
+
+        fn rec(ty: &SignatureToken) -> Vec<TypeParameterIndex> {
+            match ty {
+                Bool | Address | U64 | String | ByteArray => vec![],
+                TypeParameter(idx) => vec![*idx],
+                Reference(ty) | MutableReference(ty) => rec(ty),
+                Struct(_, tys) => tys.iter().map(rec).flatten().collect(),
+            }
+        }
+
+        let mut type_params = rec(ty);
+        type_params.sort();
+        type_params.dedup();
+        type_params
+    }
+
+    /// Builds a graph G such that
+    ///   - Each type formal of a generic function is a node in G.
+    ///   - There is an edge from type formal f_T to g_T if f_T is used to instantiate g_T in a
+    ///     call.
+    ///     - Each edge is labeled either `Identity` or `TyConApp`. See `Edge` for details.
+    fn build_graph(&mut self) {
+        for def in self
+            .module
+            .function_defs()
+            .iter()
+            .filter(|def| !def.is_native())
+        {
+            for instr in &def.code.code {
+                if let Bytecode::Call(other_func_idx, type_actuals_idx) = instr {
+                    let other_func_handle = self.module.function_handle_at(*other_func_idx);
+                    if other_func_handle.module != ModuleHandleIndex::new(0) {
```

Comment:
It is fine to use SELF, especially in preference to the number 0.  You should not sacrifice performance.

Reply:
There are other reasons `FunctionDefinitionIndex` is preferred. I mean it should have been used in the first place instead of `FunctionHandleIndex`, as such violations can only appear in function definitions (no cross module mutual recursions), and it's better we have `FunctionDefinitionIndex` in the error messages.